src/main/java/net/neoforged/neoforge/attachment/AttachmentHolder.java

```diff
@@ -95,7 +111,7 @@ public final CompoundTag serializeAttachments() {
         CompoundTag tag = null;
         for (var entry : attachments.entrySet()) {
             var type = entry.getKey();
-            if (type.serializer != null) {
+            if (type.serializer != null && !(((Predicate<Object>) type.skipSerialization).test(entry.getValue()))) {
```

Comment:
Another option is to let the serializer return null to skip serialization. Would that be preferable? Might not work nicely with codecs...

Reply:
I think it's better to return null to skip serialization, yeah. We can add a helper to pass both a codec and a shouldSerialize predicate.

Reply:
So change `IAttachmentSerializer#write` be `@Nullable`? And then add two overloads for `AttachmentType#serializable`, one for `AttachmentType.Builder#serialize(AttachmentSerializer)` and one for `AttachmentType.Builder#serialize(Codec)` that also accepts a predicate?

Reply:
I'd only do the Codec overload but yes. (INBTSerializable can just return null)

Reply:
If so should the one that takes an INBTSerializable even have a predicate passed or do we want to just say people should have their serializable object return null in those cases?

Reply:
Ignore me github didn't load your response.

Reply:
Modified how it is done some to have it be nullable, which also comes with not having it get (needlessly) copied to a new object when a copy takes place.